### PR TITLE
Refactor atoms

### PR DESCRIPTION
**Pull request by talekhinezh**
* Implement RadixEngine's RadixEngineAtom with consensus specific LedgerAtom
* Move most middleware classes from RadixEngine to core
* Use Atom only for api side (deprecate in the future)

### CODE REVIEW

File: radixdlt/src/main/java/com/radixdlt/middleware2/LedgerAtom.java [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+			super(error);
+			this.dataPointer = dataPointer;
+		}
+
+		public DataPointer getDataPointer() {
+			return dataPointer;
+		}
+	}
+
+	static ImmutableList<CMMicroInstruction> toCMMicroInstructions(List<ParticleGroup> particleGroups) throws LedgerAtomConversionException {
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/middleware2/LedgerAtom.java:153]: This has been copied over, correct?

File: radixdlt/src/main/java/com/radixdlt/middleware2/PowFeeComputer.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+package com.radixdlt.middleware2;
+
+import com.radixdlt.crypto.Hash;
+import com.radixdlt.universe.Universe;
+import com.radixdlt.utils.POW;
+import java.util.function.Supplier;
+
+/**
+ * Temporary class for computing pow spent in an atom
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/middleware2/PowFeeComputer.java]: Temporary because?

File: radixdlt/src/main/java/com/radixdlt/middleware2/converters/AtomConversionException.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * either express or implied.  See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package com.radixdlt.middleware2.converters;
+
+import com.radixdlt.constraintmachine.DataPointer;
+import java.util.Objects;
+
+public final class AtomConversionException extends Exception {
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/middleware2/converters/AtomConversionException.java:27]: Document me :)

File: radixdlt/src/test/java/com/radixdlt/middleware2/LedgerAtomTest.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+		assertThatThrownBy(() ->
+			LedgerAtom.toCMMicroInstructions(
+				ImmutableList.of(
+					ParticleGroup.of(
+						SpunParticle.up(particle0),
+						SpunParticle.down(particle0)
+					)
+				)
+			)
+		).isInstanceOf(LedgerAtomConversionException.class);
```
Comment (flotothemoon) [inline radixdlt/src/test/java/com/radixdlt/middleware2/LedgerAtomTest.java:160]: How are we checking for two errors here?

File: radixdlt/src/test/java/com/radixdlt/middleware2/LedgerAtomTest.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+				ImmutableList.of(
+					ParticleGroup.of(
+						SpunParticle.down(particle0)
+					),
+					ParticleGroup.of(
+						SpunParticle.up(particle0)
+					)
+				)
+			)
+		).isInstanceOf(LedgerAtomConversionException.class);
```
Comment (flotothemoon) [inline radixdlt/src/test/java/com/radixdlt/middleware2/LedgerAtomTest.java:145]: Should this check be more specific to ensure that the exception really is due to the single conflict?

File: radixdlt/src/main/java/com/radixdlt/middleware2/LedgerAtomChecker.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+import com.radixdlt.atommodel.Atom;
+import com.radixdlt.crypto.Hash;
+import com.radixdlt.engine.CMSuccessHook;
+import com.radixdlt.universe.Universe;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.function.LongSupplier;
+import java.util.function.Supplier;
+
+public class LedgerAtomChecker implements CMSuccessHook<LedgerAtom> {
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/middleware2/LedgerAtomChecker.java:34]: Document me :)

File: radixdlt/src/test/java/com/radixdlt/middleware2/LedgerAtomTest.java (hunk 5 above)
Comment (flotothemoon) [reply radixdlt/src/test/java/com/radixdlt/middleware2/LedgerAtomTest.java:145]: Same for similar checks above

**Review by flotothemoon — APPROVED** [review]

File: radixdlt/src/main/java/com/radixdlt/middleware2/LedgerAtom.java (hunk 1 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/middleware2/LedgerAtom.java:153]: yup exactly

File: radixdlt/src/main/java/com/radixdlt/middleware2/PowFeeComputer.java (hunk 2 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/middleware2/PowFeeComputer.java]: gonna remove pow fees soon

File: radixdlt/src/test/java/com/radixdlt/middleware2/LedgerAtomTest.java (hunk 5 above)
Comment (talekhinezh) [reply radixdlt/src/test/java/com/radixdlt/middleware2/LedgerAtomTest.java:145]: the error types are actually deprecated as these types conflicts are no longer possible with new RE instruction set. will fix the names

File: radixdlt/src/main/java/org/radix/api/observable/AtomEventObserver.java [hunk 7]
```diff
@@ -19,8 +19,11 @@
 
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
-import com.radixdlt.identifiers.AID;
+import com.radixdlt.DefaultSerialization;
```
Comment (clstrfsck) [inline radixdlt/src/main/java/org/radix/api/observable/AtomEventObserver.java]: Not used?  Not sure why checkstyle not picking these up.  Will check.

File: radixdlt/src/main/java/org/radix/api/observable/AtomEventObserver.java [hunk 8]
```diff
@@ -19,8 +19,11 @@
...
 import com.radixdlt.atommodel.Atom;
+import com.radixdlt.identifiers.AID;
+import com.radixdlt.middleware2.LedgerAtom;
+import com.radixdlt.serialization.SerializationException;
```
Comment (clstrfsck) [inline radixdlt/src/main/java/org/radix/api/observable/AtomEventObserver.java]: Not used?

File: radixdlt/src/test/java/org/radix/serialization/LedgerAtomSerializeTest.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied.  See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package org.radix.serialization;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
```
Comment (clstrfsck) [inline radixdlt/src/test/java/org/radix/serialization/LedgerAtomSerializeTest.java]: Not used.

File: radixdlt/src/test/java/org/radix/serialization/LedgerAtomSerializeTest.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+import com.radixdlt.atommodel.Atom;
+import com.radixdlt.atommodel.message.MessageParticle;
+import com.radixdlt.constraintmachine.Spin;
+import com.radixdlt.identifiers.RadixAddress;
+import com.radixdlt.middleware2.LedgerAtom;
+import com.radixdlt.middleware2.LedgerAtom.LedgerAtomConversionException;
+import org.junit.Test;
```
Comment (clstrfsck) [inline radixdlt/src/test/java/org/radix/serialization/LedgerAtomSerializeTest.java]: Not used.

File: radixdlt/src/main/java/com/radixdlt/consensus/Vertex.java [hunk 11]
```diff
@@ -164,7 +164,7 @@ public boolean equals(Object o) {
 
 		Vertex v = (Vertex) o;
 		return Objects.equals(v.view, view)
-			&& Objects.equals(v.atom, this.atom)
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/Vertex.java:167]: I had thought we had agreed that we would be using "this"?

File: radixdlt/src/main/java/com/radixdlt/mempool/SubmissionControlImpl.java [hunk 12]
```diff
@@ -28,55 +31,70 @@
...
+				e.getDataPointer()
 			);
-			this.events.broadcast(new AtomExceptionEvent(ex, atom.getAID()));
-		} else {
-			this.mempool.addAtom(atom);
+			this.events.broadcast(new AtomExceptionEvent(e, atom.getAID()));
+			return;
 		}
+
+		this.mempool.addAtom(atom);
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/mempool/SubmissionControlImpl.java]: I think this would be better inside the try block.  That way we would avoid the awkward flow in this method.

File: radixdlt/src/main/java/com/radixdlt/mempool/SubmissionControlImpl.java [hunk 13]
```diff
@@ -28,55 +31,70 @@
...
+		} catch (AtomConversionException e) {
+			log.info(
+				"Rejecting atom {} due to conversion issues.",
+				rawAtom.getAID()
+			);
+			this.events.broadcast(new AtomExceptionEvent(e, rawAtom.getAID()));
+			return;
+		}
+
 		deserialisationCallback.accept(atom);
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/mempool/SubmissionControlImpl.java]: Would be better inside the try block as above.

File: radixdlt/src/main/java/com/radixdlt/middleware2/LedgerAtom.java [hunk 14]
```diff
@@ -0,0 +1,239 @@
+/*
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/middleware2/LedgerAtom.java:1]: Yay, header comments!

File: radixdlt/src/main/java/com/radixdlt/middleware2/LedgerAtomChecker.java [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+		this.maximumDrift = maximumDrift;
+	}
+
+	@Override
+	public Result hook(LedgerAtom atom) {
+		if (atom.getCMInstruction().getMicroInstructions().isEmpty()) {
+			return Result.error("atom has no instructions");
+		}
+
+		final boolean isMagic = Objects.equals(atom.getMetaData().get("magic"), "0xdeadbeef");
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/middleware2/LedgerAtomChecker.java:65]: Should really get rid of this.

File: radixdlt/src/main/java/com/radixdlt/middleware2/MiddlewareModule.java [hunk 16]
```diff
@@ -76,39 +78,40 @@ private ConstraintMachine buildConstraintMachine(CMAtomOS os) {
 
 	@Provides
 	@Singleton
-	private RadixEngine getRadixEngine(
+	private RadixEngine<LedgerAtom> getRadixEngine(
 			ConstraintMachine constraintMachine,
 			UnaryOperator<CMStore> virtualStoreLayer,
-			EngineStore engineStore,
+			EngineStore<LedgerAtom> engineStore,
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/middleware2/MiddlewareModule.java]: Two many tabs here.  (See what I did there?)

File: radixdlt/src/main/java/com/radixdlt/middleware2/converters/AtomToBinaryConverter.java [hunk 17]
```diff
@@ -29,19 +29,19 @@ public AtomToBinaryConverter(Serialization serializer) {
...
 	}
 
-	public Atom toAtom(byte[] ledgerEntryContent) {
+	public LedgerAtom toAtom(byte[] ledgerEntryContent) {
 		try {
-			return serializer.fromDson(ledgerEntryContent, Atom.class);
+			return serializer.fromDson(ledgerEntryContent, LedgerAtom.class);
 		} catch (SerializationException e) {
-			throw new RuntimeException("Deserialization of Atom failed");
+			throw new RuntimeException("Deserialization of Atom failed", e);
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/middleware2/converters/AtomToBinaryConverter.java]: Kudos for exception chaining, but really should be an IllegalStateException, rather than raw RuntimeException.

File: radixdlt/src/main/java/org/radix/GenerateUniverses.java [hunk 18]
```diff
@@ -164,6 +164,8 @@ private Atom createGenesisAtom(byte magic, long timestamp, long planck) throws E
 		);
 		genesisAtom.sign(universeKey);
 
+		LedgerAtom ledgerAtom = LedgerAtom.convertFromApiAtom(genesisAtom);
```
Comment (clstrfsck) [inline radixdlt/src/main/java/org/radix/GenerateUniverses.java:167]: We could just avoid the conversion and coupling by throwing a different exception.  Not really sure that ConstraintMachineValidationException is the right thing for this largely stand-alone piece of code that is unrelated to the constraint machine.

**Review by clstrfsck — APPROVED** [review]
Some unused imports to be cleaned up, and some other suggestions that I'm mostly fine with either way.  Not sure why checkstyle isn't picking up the unused imports — will check into this.

File: radixdlt/src/main/java/com/radixdlt/middleware2/LedgerAtomChecker.java (hunk 15 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/middleware2/LedgerAtomChecker.java:65]: yes!

Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/middleware2/LedgerAtomChecker.java:65]: perhaps when we implement real fees

File: radixdlt/src/main/java/org/radix/GenerateUniverses.java (hunk 18 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/org/radix/GenerateUniverses.java:167]: That's true but I think this will get resolved once we formalize atom construction in the RE.
